src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,489 @@
+	// The amount of time for which the credential is valid. Specifically, the
+	// the credential expires 'ValidTime' seconds after the 'notBefore' of the
+	// delegation certificate. The delegator shall not issue delegated
+	// credentials that are valid for more than 7 days from their current time.
+	//
+	// When this data structure is serialized, this value is converted to a
+	// uint32 representing the duration in seconds.
+	ValidTime time.Duration
+	// The signature scheme associated with the credential public key.
+	// This is expected to be the same as the CertificateVerify.algorithm
+	// sent by the server.
+	expCertVerfAlgo SignatureScheme
+	// The credential's public key.
+	PublicKey crypto.PublicKey
+}
+
+// DelegatedCredential stores the credential and its signature.
+type DelegatedCredential struct {
+	// The serialized form of the Credential
+	Raw []byte
+
+	// Credential stores the public components of a Delegated Credential
+	Cred *Credential
+
+	// The signature scheme used to sign the credential
+	Algorithm SignatureScheme
+
+	// The Credential's delegation: a signature that binds the credential to
+	// the end-entity certificate's public key
+	Signature []byte
+}
+
+// marshalSubjectPublicKeyInfo returns a DER encoded SubjectPublicKeyInfo structure
+// (as defined in the X.509 standard) for the credential.
+// TODO: maybe we can move this as well to x509
+// TODO: add the other signatures flavors, as defined in common.go
+func (cred *Credential) marshalPublicKeyInfo() ([]byte, error) {
+	switch cred.expCertVerfAlgo {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		serPub, err := x509.MarshalPKIXPublicKey(cred.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		return serPub, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported signature scheme: 0x%04x", cred.expCertVerfAlgo)
+	}
+}
+
+// unmarshalPublicKeyInfo parses a DER encoded PublicKeyInfo
+// structure into a public key and its corresponding algorithm.
+// TODO: add the other signatures flavors, as defined in common.go
+func unmarshalPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+	pubKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := pubKey.(type) {
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("curve %s s not supported", curveName)
+		}
+
+	default:
+		return nil, 0, fmt.Errorf("unsupported delgation key type: %T", pk)
+	}
+}
+
+// marshal encodes the credential struct of the Delegated Credential.
+func (cred *Credential) marshal() ([]byte, error) {
+	ser := make([]byte, 4)
+	binary.BigEndian.PutUint32(ser, uint32(cred.ValidTime/time.Second))
```

Comment:
What are the units of `ValidTime`?

Reply:
`ValidTime` is `time.Duration`, which is an int64.

Reply:
👍